src/wp-includes/default-filters.php

```diff
@@ -596,9 +596,11 @@
 add_filter( 'block_editor_settings_all', 'wp_add_editor_classic_theme_styles' );
 
 // Global styles can be enqueued in both the header and the footer. See https://core.trac.wordpress.org/ticket/53494.
-add_action( 'wp_enqueue_scripts', 'wp_enqueue_global_styles' );
+add_action( 'init', 'wp_enqueue_global_styles', 1 );
```

Comment:
I changed this to `init` as it seems to be the only way to get it to run before the [core block styles are registered](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/blocks/index.php#L141)

Reply:
cc @scruffian @ajlende @aristath as they may have the most experience with all the scenarios, IIRC.

I couldn't experiment/test this change because my wordpress-develop env wouldn't start up, so I'm trying to recall from memory / skimming through the code. There are two scenarios: the core blocks styles are loaded with a separate stylesheet or bundled as part of the block library.

As I understand it, the order of styles after this change would be

## Single stylesheet

Before:

1. `wp-block-library-css`
2. `global-styles-inline-css` (both top-level & block-level CSS)
3. `theme-style-css`

After

1. `global-styles-inline-css` (both top-level & block-level CSS)
2. `wp-block-library-css`
3. `theme-style-css`

## Separate stylesheets

Before:

1. `wp-block-blockname-inline-css` (the block styles + the block-level CSS coming from global styles)
2.  `wp-block-library-inline-css`
3. `global-styles-inline-css` (only top-level CSS)
4. `core-block-supports-inline-css`

After:

1. `global-styles-inline-css` (only top-level CSS)
2. `wp-block-blockname-inline-css` (the block styles + the block-level CSS coming from global styles)
3.  `wp-block-library-inline-css`
4. `core-block-supports-inline-css`

Is this the goal / what happens?

---

Notes:

- There are some block styles that are absorbed as part of the global styles. I've found some for [navigation](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/block.json#L127) and [pullquote](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/pullquote/block.json#L67). These are enqueued where the block-level global styles are.
- In testing how it works now using a different env, I've noticed that even in the case of "separate stylesheets", the `global-styles-inline-css` stylesheet has a lot of block styles (tested with TwentyTwentyFour). I may be a bit rusty on how this all is supposed to work, but it sounds like a bug to me. I've found styles for `.wp-block-calendar`, `.wp-block-categories`, `.wp-block-post-comments-form`, `.wp-block-loginout`, `.wp-block-post-terms`, `.wp-block-query-title`, `.wp-block-quote`, `.wp-block-search`, `.wp-block-separator`. Some of them do not have style rules, for example, `.wp-block-separator {}`.

Reply:
The change LGTM to me. Of course "global" styles should always be loaded before "local" styles :) 
Was just wondering (nitpick) if using `wp_enqueue_scripts` priority 1 would be sufficient. As we are enqueueing stylesheets using the appropriate hook would be better for "educational purposes" :)

Reply:
Thanks for looking into this!

> Was just wondering (nitpick) if using wp_enqueue_scripts priority 1 would be sufficient

I tried and it didn't work, I think because `wp_enqueue_scripts` always runs after `init`.

@oandregal the separate stylesheets scenario you outline is correct; I'll need to double-check the single stylesheet but I think it's also right.

Reply:
> The change LGTM. Of course "global" styles should always be loaded before "local" styles :)

This is incorrect. It's important to me that we all share the same context. This is how it works now:

1. block-level styles provided by the block (the block stylesheet)
2. global styles (top or block-level) provided by core/blocks/theme/user (different theme.json files & user via global styles sidebar)
3. block-level styles provided by the user (what we should call "local", these are serialized in the post as classes/style, and they should have the maximum priority)

My understanding is that we want to do this instead:

1. global styles (top level) provided by core/blocks/theme/user
2. block-level styles provided by the block (the block stylesheet)
3. global styles (block-level) provided by core/blocks/theme/user
4. block-level styles provided by the user

The top-level global styles should be: presets + some styles that are attached to the `body`.

I mentioned above that I see still some "block styles" loaded in the "separate stylesheet" scenario (my understanding this is a bug). With this change, they'll be enqueued in 1, and can create conflicts because they should be enqueued in 3.

Reply:
> it lacks a bit of nuance. Styles is a very loaded domain, the same words mean different things to people

Yeaaa, can see that too :) 

In my mind "global styles" means styles that apply to all similar elements, for example blocks, or image blocks, or paragraph blocks, etc. "Local styles" means styles that apply to only one of many elements. For example only to one image block when there are 5 on the page, etc.

Thinking it would be great to add some explanations in comments at the top of all CSS files in an attempt to get everybody on the same page. Can also be added in docblocks where the CSS is outputted from JS or PHP.

Reply:
For reference, this is [the documentation](https://developer.wordpress.org/block-editor/explanations/architecture/styles/) we have for styles (naming, how they work, etc.).

Reply:
OK I looked at this carefully and can confirm the following order of styles:

## Single stylesheet

### trunk

* `wp-block-library-css` is loaded as an external stylesheet
* `wp-block-library-inline-css` 
* `global-styles-inline-css` containing base global styles followed by block-specific global styles
* `core-block-supports-inline-css`

### this PR

* `global-styles-inline-css` containing base global styles followed by block-specific global styles
* `wp-block-library-css` is loaded as an external stylesheet
* `wp-block-library-inline-css` 
* `core-block-supports-inline-css`

## Separate stylesheets

### trunk

* `wp-block-[blockname]-inline-css` containing _both_ the block library styles for that block _and_ the block-specific global styles
* in between the block style tags, external stylesheets are loaded for some blocks (e.g. I've seen Navigation, Image and Button, seems they change depending on the combination of blocks on the page)
* `wp-block-library-inline-css` (sometimes this is replaced by a block library "common" external stylesheet, no idea why)
* `global-styles-inline-css` containing only base global styles
* `core-block-supports-inline-css`

## this PR

* `global-styles-inline-css` containing only base global styles
*  in between style tags, external stylesheets for some blocks
* `wp-block-library-inline-css` (sometimes this is replaced by a block library "common" external stylesheet, no idea why)
* `core-block-supports-inline-css`

So your assessment is essentially correct @oandregal. 

I think the only suboptimal change in this PR is in the single stylesheet scenario, where block library CSS is now loaded _after_ block-specific global styles. It should be the other way around, but ideally base global styles would _also_ be added before block library styles. I'll look into splitting base and block-specific global styles into separate style tags as a workaround.

> There are some block styles that are absorbed as part of the global styles. I've found some for [navigation](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/block.json#L127) and [pullquote](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/pullquote/block.json#L67).

I'm probably missing something, but I don't see those styles appearing at all either on trunk or this PR branch. Tested by adding a Pullquote block to a page and checking that a line-height of `1.6` doesn't appear anywhere in the styles cascade.

> I've noticed that even in the case of "separate stylesheets", the global-styles-inline-css stylesheet has a lot of block styles

Those extra block styles added to `global-styles-inline-css` are coming from pre-block custom css defined in the theme's theme.json, e.g. [here](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentyfour/theme.json#L727). TT4 has a bunch of instances.

> with this change, they'll be enqueued in 1, and can create conflicts because they should be enqueued in 3

Good point. You mean they should be enqueued _after_ the block library styles, right? I'll look into it.

Reply:
> Those extra block styles added to global-styles-inline-css are coming from per-block custom css defined in the theme's theme.json, e.g. [here](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-content/themes/twentytwentyfour/theme.json#L727). TT4 has a bunch of instances.

Oh, thanks for sharing. I think this is a bug: block styles should always be together. Why `styles.block/search.css` should be processed differently than `styles.block/search.typography`?